### ISSUE TITLE
TabularSHAP base value varies across instances

### ISSUE THREAD

**Issue by boltzmann-brain** [microsoft/SynapseML#1199]
**Describe the bug**
Based on [this sample notebook](https://github.com/microsoft/SynapseML/blob/master/notebooks/Interpretability%20-%20Tabular%20SHAP%20explainer.ipynb), it sounds like the 'base' value calculated by TabularSHAP is the average score in the background dataset. Since the background dataset is fixed, I thought this meant that the base value will be the same across all instances being explained. However, in practice it varies very slightly (somewhere in the 3rd-6th significant figure or so) across instances.

**To Reproduce**
Run the TabularSHAP example in [the sample notebook](https://github.com/microsoft/SynapseML/blob/master/notebooks/Interpretability%20-%20Tabular%20SHAP%20explainer.ipynb). The instances will have slightly different base values. For example, when I run it the first five instances have the following base values:
`0.2035667`, `0.2035668`, `0.20356692`, `0.20356715`, `0.20356685`.

In practice these slight variations don't affect what we would do with the explanation, but I want to make sure that they are expected and not due to a bug. When I tried on a different (proprietary) dataset and model, I got variations that were larger than those found in the above tutorial, affecting the third significant figure. Example values:
`0.22217013`, `0.22328514`.

**Expected behavior**

If the base value is calculated from the average score of the background dataset, then it should be identical across instances. If it's not calculated from the average score of the background dataset, it would be great to clarify how it's calculated so that we can be reassured that the variations are not due to a bug.

**Info (please complete the following information):**
 - MMLSpark Version: 1.0.0-rc3-148-87ec5f74-SNAPSHOT
 - Spark Version 3.0.1
 - Spark Platform: Databricks 7.3 LTS

@memoryz

**Comment by memoryz** [comment]
@boltzmann-brain , thanks for trying it out!
TLDR: very small variations in base values are expected.

The base value is calculated this way: for each observation to be explained, we create a bunch of simulated coalitions. One of the coalition is all zeros (simulating none of the features is known to the model), and one of the coalition is all ones (simulating all of the features are known to the model), and the rest are sampled based on SHAP kernel. When we build the WLS regression, we assign a very big weight (default value 1E8) to the all-zeros coalition and the all-ones coalition to enforce two constraints: the base value should be close to the mean output from the background dataset; the SHAP values add up to the output of the observation.

Usually the big weight should give you a base value very close to the mean output from the background dataset (~1E-5 difference). If that's not the case, try increase numSamples, or increase the size of background dataset, and see if it makes a difference.

**Comment by boltzmann-brain** [comment]
@memoryz Thanks, that makes sense. I'll close the issue.
